eggtimer.js

```
@@ -1,326 +1,681 @@
+            prParams.number = prContext.pr.number.toString();
+            let prPromise = updatePR(prParams);
+
+            let addLabelParams = commonParams();
+            addLabelParams.number = prContext.pr.number.toString();
+            addLabelParams.labels = [];
+            addLabelParams.labels.push(MergedLabel);
+            let addLabelPromise = addLabels(addLabelParams);
+
+            let deleteTagParams = commonParams();
+            deleteTagParams.ref = mergingTag(prContext.pr.number);
+            let tagPromise = deleteReference(deleteTagParams);
+
+            let delLabelParams = commonParams();
+            delLabelParams.number = prContext.pr.number.toString();
+            delLabelParams.name = MergeFailedLabel;
+            let delLabelPromise = removeLabel(delLabelParams);
+
+            return Promise.all([prPromise, addLabelPromise, tagPromise, delLabelPromise]);
+        })
+        .then((results) => {
+            const state = results[0];
+            const labelsUpdated = results[1];
+            if (state === "closed" && labelsUpdated)
+                return Promise.resolve(true);
+            return Promise.reject("cleanup failed");
+        })
+        .then((result) => {
+             assert(result === true);
+             processNextPR();
+             return Promise.resolve(true);
+         })
+        .catch((err) => {
+            if (errorLabel === null) {
+                return Promise.resolve(true);
+            } else {
+               console.error("Error merging auto_branch(" + prContext.autoSha + ") into master:", err);
+               processNextPR();
+               let params = commonParams();
+               params.number = prContext.pr.number.toString();
+               params.labels = [];
+               params.labels.push(errorLabel);
+               return addLabels(params);
+            }
+        })
+        .then((result) => {
+             assert(result === true);
+        })
+        .catch((err) => {
```

It feels like there are too many error handling (.catch) places in the code, which probably also means that there are not enough (i.e., some cases are missed). AFAICT, there should be no need to catch all exceptions close to their origin. Is it possible to unify handling of all exceptions at the top (i.e., processEvent() or similar) level?

---

```
Is it possible to unify handling of all exceptions at the top (i.e., processEvent() or similar) level?
```

I was thinking about that: in current implementation this is possible if convert 'startMerging()' and 'checkMergePreconditions()' into promises and make a high-level try-catch. I doubt, however, that this would be a valuable optimization. finishMerging() is handled separately: it is activated if a auto_status has come.

---

> finishMerging() is handled separately: it is activated if a auto_status has come

As we discussed, this will not be true in the next revision. Perhaps that will help you centralize error handling. Still, there are different _kinds_ of errors (with different effect on GitHub and bot states) so error handling will not be trivial even if it is centralized.